line_profiler/_line_profiler.pyx

```diff
@@ -1462,33 +1460,28 @@ cdef inline inner_trace_callback(
             time = hpTimer()
             has_time = True
         ident = PyThread_get_thread_ident()
-        last_map = &(prof._c_last_time[ident])
         # deref() is Cython's version of the -> accessor in C++. if we don't use deref then
         # Cython thinks that when we index last_map,
         # we want pointer indexing (which is not the case)
-        if deref(last_map).count(block_hash):
-            old = deref(last_map)[block_hash]
+        last_map = &(prof._c_last_time[ident])
+        if last_contains(last_map, block_hash):
+            # Read once; no extra insertion
+            old = last_get_value(last_map, block_hash)
             line_entries = &(prof._c_code_map[code_hash])
-            key = old.f_lineno
-            if not deref(line_entries).count(key):
-                deref(line_entries)[key] = LineTime(code_hash, key, 0, 0)
-            entry = &(deref(line_entries)[key])
-            # Note: explicitly `deref()`-ing here causes the new values
-            # to be assigned to a temp var;
-            # meanwhile, directly dot-accessing a pointer causes Cython
-            # to correctly write `ptr->attr = (ptr->attr + incr)`
+            entry = line_ensure_entry(line_entries, old.f_lineno, code_hash)
             entry.nhits += 1
             entry.total_time += time - old.time
+            have_last = True
+        else:
+            have_last = False
+
         if is_line_event:
-            # Get the time again. This way, we don't record much time
-            # wasted in this function.
-            deref(last_map)[block_hash] = LastTime(lineno, hpTimer())
-        elif deref(last_map).count(block_hash):
-            # We are returning from a function, not executing a line.
-            # Delete the last_time record. It may have already been
-            # deleted if we are profiling a generator that is being
-            # pumped past its end.
-            deref(last_map).erase(deref(last_map).find(block_hash))
+            # Record fresh timestamp for this block
+            last_set_now(last_map, block_hash, lineno)
+        elif have_last:
```

Comment:
Are you absolutely sure that this won't cause any test issues or incorrectness? Seems iffy to me, but if it passes tests then I guess it'll be fine.

Reply:
No, I'm not absolutely sure. I don't have every C++ memory semantic memorized. It seemed ok to me. 

If this PR is more trouble than its worth, we can just accept your version.  I'm comfortable enough with the logic at least.

Reply:
I'll definitely include as much of your version as I can without breaking any tests or slowing down the code. Gimme an hour or so and, barring unforeseen test failures, I'll push a new version to this branch that should take in as much of your code as will work!

Reply:
Ok, can confirm it does pass tests!